File: src/main/java/archi/pole/rest/PremierREST.java
package archi.pole.rest;

import io.vertx.core.AbstractVerticle;
import io.vertx.core.Future;
//import io.vertx.core.Vertx;
import io.vertx.core.json.JsonArray;
import io.vertx.core.json.JsonObject;
import io.vertx.ext.web.Router;
import io.vertx.ext.web.RoutingContext;
import io.vertx.ext.web.handler.BodyHandler;

import java.util.HashMap;
import java.util.Map;


/**
 * Created by johann on 20/03/17.
 */

public class PremierREST extends AbstractVerticle {

    private Map<String, JsonObject> products = new HashMap<>();

    @Override
    public void start(Future fut) {
        System.out.println("------------------------Démarrage de l'instance Verticle!");
        setUpInitialData();
        Router router = Router.router(vertx);
        router.route().handler(BodyHandler.create());
        router.get("/products").handler(this::handleListProducts);
        router.get("/stop").handler(this::stopVertx);
        vertx.createHttpServer().requestHandler(router::accept).listen(8080, res -> fut.complete());
    }

    private void stopVertx(RoutingContext routingContext) {
        routingContext.response().putHeader("content-type", "text/html").end("Instance Verticle stoppée");
        vertx.close();
    }

    private void handleListProducts(RoutingContext routingContext) {
        JsonArray arr = new JsonArray();
        products.forEach((k, v) -> arr.add(v));
        routingContext.response().putHeader("content-type", "application/json").end(arr.encodePrettily());
    }

    private void setUpInitialData() {
        addProduct(new JsonObject().put("id", "prod3568").put("name", "Egg Whisk").put("price", 3.99).put("weight", 150));
        addProduct(new JsonObject().put("id", "prod7340").put("name", "Tea Cosy").put("price", 5.99).put("weight", 100));
        addProduct(new JsonObject().put("id", "prod8643").put("name", "Spatula").put("price", 1.00).put("weight", 80));
    }

    private void addProduct(JsonObject product) {
        products.put(product.getString("id"), product);
    }

    @Override
    public void stop() {
        System.out.println("----------------------Verticle stoppé!");
    }
}

File: src/main/java/archi/pole/rest/PremierREST.kt
package archi.pole.rest

import io.vertx.core.AbstractVerticle
import io.vertx.core.Future
import io.vertx.core.Handler
import io.vertx.core.http.HttpServerResponse
import io.vertx.core.json.Json
import io.vertx.core.json.JsonArray
import io.vertx.core.json.JsonObject
import io.vertx.ext.web.Router
import io.vertx.ext.web.RoutingContext
import io.vertx.ext.web.handler.BodyHandler


import io.vertx.ext.mongo.MongoClient

/**
 * Created by johann on 20/03/17.
 */

@Suppress("unused")
class PremierREST : AbstractVerticle() {
    lateinit var client: MongoClient

    override fun start(fut: Future<Void>) {
        val router = Router.router(vertx)
        router.route().handler(BodyHandler.create())
        router.get("/companies").handler(handleCompanies)
        router.get("/company/:companyid").handler(handleCompany)

        vertx.createHttpServer().requestHandler({ router.accept(it) }).listen(8080, { res -> fut.complete() })

        client = Connection().init(vertx)

        Mock().setupInitialData(client)
    }

    val handleCompanies = Handler <RoutingContext> { req ->
        getCompanies(req)
    }

    val handleCompany = Handler <RoutingContext> { req ->
        var id = req.request().getParam("companyid")
        getCompany(req, id)
    }

    fun getCompanies(req: RoutingContext) {
        client.find("companies", JsonObject(), { res ->
            if (res.succeeded()) {
                var companies = JsonArray()
                for (company in res.result()) {
                    val consultantsJson = company.getJsonArray("consultants")
                    val consultants = mutableListOf<Consultant>()
                    for ((index, consultant) in consultantsJson.withIndex()) {
                        //TODO: Smart casts de Kotlin: faire confirmer
                        if (consultant is JsonObject) {
                            consultants.add(index, Consultant(consultant.getString("name"), consultant.getString("forename")))
                        }
                    }
                    companies.add(JsonObject.mapFrom(Company(company.getString("name"), company.getString("address"), consultants)))
                }
                req.response().endWithJson(companies)
            } else {
                res.cause().printStackTrace()
            }
        })
    }

    /**
     * Get company from name
     */
    fun getCompany(req: RoutingContext, id: String) {
        client.find("companies", JsonObject(), { res ->
            if (res.succeeded()) {
                var foundCompany = Company()
                for (company in res.result()) {
                    if (company is JsonObject) {
                        if (company.getString("name").equals(id)) {
                            val consultantsJson = company.getJsonArray("consultants")
                            val consultants = mutableListOf<Consultant>()
                            for ((index, consultant) in consultantsJson.withIndex()) {
                                if (consultant is JsonObject) {
                                    consultants.add(index, Consultant(consultant.getString("name"), consultant.getString("forename")))
                                }
                                foundCompany = Company(company.getString("name"), company.getString("address"), consultants)
                            }
                        }
                    }
                }

                if (foundCompany.name != "") {
                    req.response().endWithJson(JsonObject().put("result", foundCompany.toJsonObject()))
                } else {
                    req.response().endWithJson(JsonObject().put("result", "Company not found"))
                }

            }
        })
    }

    fun HttpServerResponse.endWithJson(obj: Any) {
        this.putHeader("Content-Type", "application/json; charset=utf-8").end(Json.encodePrettily(obj))
    }


}